### PR TITLE
hosts: adding domain and realm variables

### PR DESCRIPTION
**Pull request by danlavu**
- adding a generic domain and realm variable that can be used in tests, as provider.host.domain/realm
 - updated existing functions to find the domain to use the new variable instead

### CODE REVIEW

**Comment by danlavu** [comment]
I know _variable is supposed to be used for temporary variables, its shorter than domain_str/string, are you guys okay with this?

**Comment by spoore1** [comment]
FYI, at least some of the tox failures are fixed in PR #39 with this commit:

https://github.com/SSSD/sssd-test-framework/pull/39/commits/a270cd1b7655f0cce5ff10c32f0bf1c815f54a43

Asking in that PR if I should break out my commit into a new/separate PR to resolve this.

**Comment by spoore1** [comment]
Ok, PR #75 was merged to fix the black reformatting.  I think you can rebase this now to fix the tox 3.10 tests.

**Review by spoore1 — APPROVED** [review]
LGTM

File: sssd_test_framework/hosts/base.py [hunk 1]
```diff
@@ -134,6 +134,12 @@ def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
         self.client: dict[str, Any] = self.config.get("client", {})
 
+        self.domain_str: str = self.config.get("domain", "test")
```
Comment (pbrezina) [inline sssd_test_framework/hosts/base.py:137]: Is it really a domain of directory server or a Kerberos domain?

Comment (danlavu) [reply sssd_test_framework/hosts/base.py:137]: It's the domain. 

https://github.com/SSSD/sssd/pull/7156/files

**Comment by pbrezina** [comment]
> I know _variable is supposed to be used for temporary variables, its shorter than domain_str/string, are you guys okay with this?

`_variable` is mainly used for private stuff, something that should not be used outside the class. There is also `__variable` that enforces it, it can not be accessed even in child classes.

I merged https://github.com/next-actions/pytest-mh/pull/39 that removed the name conflict so it is possible to use `domain` and `realm`.

**Comment by pbrezina** [comment]
Domain should filled in IPA, AD and Samba host from `ipa_domain` and `ad_domain` and realm should default to uppercase of the domain.

**Comment by danlavu** [comment]
> > I know _variable is supposed to be used for temporary variables, its shorter than domain_str/string, are you guys okay with this?
> 
> `_variable` is mainly used for private stuff, something that should not be used outside the class. There is also `__variable` that enforces it, it can not be accessed even in child classes.
> 
> I merged [next-actions/pytest-mh#39](https://github.com/next-actions/pytest-mh/pull/39) that removed the name conflict so it is possible to use `domain` and `realm`.

That's great, domain and realm is a lot cleaner.

**Comment by danlavu** [comment]
> Domain should be filled in IPA, AD and Samba host from `ipa_domain` and `ad_domain` and realm should default to uppercase of the domain.

Updated.

**Review by pbrezina — CHANGES_REQUESTED** [review]
Hi, please, apply the following diff.

It makes it more generic, function for every domain.

```diff
diff --git a/sssd_test_framework/hosts/ad.py b/sssd_test_framework/hosts/ad.py
index 9b8d895..d6cbab1 100644
--- a/sssd_test_framework/hosts/ad.py
+++ b/sssd_test_framework/hosts/ad.py
@@ -35,17 +35,20 @@ class ADHost(BaseDomainHost):
 
         self._features: dict[str, bool] | None = None
 
-        self.ad_domain: str = self.client["ad_domain"]
-        """
-        Active Directory domain name.
-        """
-
         # Additional client configuration
         self.client.setdefault("id_provider", "ad")
         self.client.setdefault("access_provider", "ad")
         self.client.setdefault("ad_server", self.hostname)
         self.client.setdefault("dyndns_update", False)
 
+        # Use different default for domain
+        if "domain" not in self.config and "ad_domain" in self.client:
+            self.domain = self.client["ad_domain"]
+
+        # Use different default for realm
+        if "realm" not in self.config:
+            self.realm = self.domain.upper()
+
         # Lazy properties
         self.__naming_context: str | None = None
 
diff --git a/sssd_test_framework/hosts/base.py b/sssd_test_framework/hosts/base.py
index 2b4a540..e1fc1e3 100644
--- a/sssd_test_framework/hosts/base.py
+++ b/sssd_test_framework/hosts/base.py
@@ -134,15 +134,11 @@ class BaseDomainHost(BaseBackupHost):
         super().__init__(*args, **kwargs)
         self.client: dict[str, Any] = self.config.get("client", {})
 
-        if self.role == "ad" or self.role == "samba":
-            self.domain: str = self.client.get("ad_domain")
-            self.realm: str = self.domain.upper()
-        if self.role == "ipa":
-            self.domain: str | MultihostHost = self.client.get("ipa_domain")
-            self.realm: str = self.domain.upper()
-        """
-        Directory domain and realm names.
-        """
+        self.domain: str = self.config.get("domain", "test")
+        """Identity domain name."""
+
+        self.realm: str = self.config.get("realm", self.domain.upper())
+        """Kerberos realm."""
 
 
 class BaseLDAPDomainHost(BaseDomainHost):
diff --git a/sssd_test_framework/hosts/ipa.py b/sssd_test_framework/hosts/ipa.py
index 3d00c6a..49ddb08 100644
--- a/sssd_test_framework/hosts/ipa.py
+++ b/sssd_test_framework/hosts/ipa.py
@@ -55,6 +55,14 @@ class IPAHost(BaseDomainHost):
         self.client.setdefault("ipa_server", self.hostname)
         self.client.setdefault("dyndns_update", False)
 
+        # Use different default for domain
+        if "domain" not in self.config and "ipa_domain" in self.client:
+            self.domain = self.client["ipa_domain"]
+
+        # Use different default for realm
+        if "realm" not in self.config:
+            self.realm = self.domain.upper()
+
     @property
     def features(self) -> dict[str, bool]:
         """
diff --git a/sssd_test_framework/hosts/kdc.py b/sssd_test_framework/hosts/kdc.py
index c125648..e2bc60c 100644
--- a/sssd_test_framework/hosts/kdc.py
+++ b/sssd_test_framework/hosts/kdc.py
@@ -40,9 +40,6 @@ class KDCHost(BaseDomainHost):
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
 
-        self.realm: str = self.config.get("realm", "TEST")
-        self.krbdomain: str = self.config.get("domain", "test")
-
         self.client["auth_provider"] = "krb5"
 
     def backup(self) -> None:
diff --git a/sssd_test_framework/hosts/samba.py b/sssd_test_framework/hosts/samba.py
index 9f94520..6a5416c 100644
--- a/sssd_test_framework/hosts/samba.py
+++ b/sssd_test_framework/hosts/samba.py
@@ -25,17 +25,20 @@ class SambaHost(BaseLDAPDomainHost):
 
         self._features: dict[str, bool] | None = None
 
-        self.ad_domain: str = self.client["ad_domain"]
-        """
-        Active Directory domain name.
-        """
-
         # Additional client configuration
         self.client.setdefault("id_provider", "ad")
         self.client.setdefault("access_provider", "ad")
         self.client.setdefault("ad_server", self.hostname)
         self.client.setdefault("dyndns_update", False)
 
+        # Use different default for domain
+        if "domain" not in self.config and "ad_domain" in self.client:
+            self.domain = self.client["ad_domain"]
+
+        # Use different default for realm
+        if "realm" not in self.config:
+            self.realm = self.domain.upper()
+
     @property
     def features(self) -> dict[str, bool]:
         """
diff --git a/sssd_test_framework/roles/ad.py b/sssd_test_framework/roles/ad.py
index f1cda13..a680a81 100644
--- a/sssd_test_framework/roles/ad.py
+++ b/sssd_test_framework/roles/ad.py
@@ -55,7 +55,7 @@ class AD(BaseWindowsRole[ADHost]):
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
 
-        self.domain: str = self.host.ad_domain
+        self.domain: str = self.host.domain
         """
         Active Directory domain name.
         """
@@ -623,7 +623,7 @@ class ADUser(ADObject):
             "OtherAttributes": (self.cli.option.PLAIN, self._attrs_to_hash(unix_attrs)),
             "Enabled": (self.cli.option.PLAIN, "$True"),
             "Path": (self.cli.option.VALUE, self.path),
-            "EmailAddress": (self.cli.option.PLAIN, f'{self.name}@{self.host.client["ad_domain"]}'),
+            "EmailAddress": (self.cli.option.PLAIN, f'{self.name}@{self.host.domain}'),
             "GivenName": (self.cli.option.PLAIN, "dummyfirstname"),
             "Surname": (self.cli.option.PLAIN, "dummylastname"),
         }
diff --git a/sssd_test_framework/roles/kdc.py b/sssd_test_framework/roles/kdc.py
index 3106063..eae2f9e 100644
--- a/sssd_test_framework/roles/kdc.py
+++ b/sssd_test_framework/roles/kdc.py
@@ -142,8 +142,8 @@ class KDC(BaseLinuxRole[KDCHost]):
             }}
 
             [domain_realm]
-            .{self.host.krbdomain} = {self.host.realm}
-            {self.host.krbdomain} = {self.host.realm}
+            .{self.host.domain} = {self.host.realm}
+            {self.host.domain} = {self.host.realm}
         """
         ).lstrip()
 
diff --git a/sssd_test_framework/roles/samba.py b/sssd_test_framework/roles/samba.py
index 15e45d7..31ef452 100644
--- a/sssd_test_framework/roles/samba.py
+++ b/sssd_test_framework/roles/samba.py
@@ -51,7 +51,7 @@ class Samba(BaseLinuxLDAPRole[SambaHost]):
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
 
-        self.domain: str = self.host.ad_domain
+        self.domain: str = self.host.domain
         """
         Active Directory domain name.
         """
```

**Comment by danlavu** [comment]
Updated this with your patch and updated the commit message.

**Review by pbrezina — APPROVED** [review]

**Pull request merged**